src/mvba/abba/mod.rs

```diff
@@ -1,56 +1,567 @@
+            Action::PreVote(action) => {
+                let pre_votes = self.get_mut_pre_votes_by_round(action.round);
+                if pre_votes.contains_key(sender) {
+                    return Ok(false);
+                }
+
+                pre_votes.insert(*sender, action.clone());
+            }
+            Action::MainVote(action) => {
+                let main_votes = self.get_mut_main_votes_by_round(action.round);
+                if main_votes.contains_key(sender) {
+                    return Ok(false);
+                }
+
+                main_votes.insert(*sender, action.clone());
+            }
+        }
+        Ok(true)
+    }
+
+    fn check_message(&mut self, sender: &NodeId, msg: &Message) -> Result<()> {
+        if msg.id != self.id {
+            return Err(Error::InvalidMessage(format!(
+                "invalid ID. expected: {}, got {}",
+                self.id, msg.id
+            )));
+        }
+
+        match &msg.action {
+            Action::PreVote(action) => {
+                // check the validity of the S-signature share on message (ID, pre-vote, r, b)
+                let sign_bytes = self.pre_vote_bytes_to_sign(action.round, &action.value)?;
+                if !self
+                    .pub_key_set
+                    .public_key_share(sender)
+                    .verify(&action.sig_share, &sign_bytes)
+                {
+                    return Err(Error::InvalidMessage("invalid signature share".to_string()));
+                }
+
+                match &action.justification {
+                    PreVoteJustification::RoundOneNoJustification => {
+                        if action.round != 1 {
+                            return Err(Error::InvalidMessage(format!(
+                                "invalid round. expected 1, got {}",
+                                action.round
+                            )));
+                        }
+                    }
+                    PreVoteJustification::RoundOneJustification(subject, proof) => {
+                        if action.round != 1 {
+                            return Err(Error::InvalidMessage(format!(
+                                "invalid round. expected 1, got {}",
+                                action.round
+                            )));
+                        }
+
+                        if subject != &self.subject {
+                            return Err(Error::InvalidMessage(format!(
+                                "invalid subject. expected {}, got {}",
+                                self.subject, subject
+                            )));
+                        }
+
+                        if !self
+                            .pub_key_set
+                            .public_key()
+                            .verify(proof, &subject.to_bytes())
+                        {
+                            return Err(Error::InvalidMessage("invalid proof".to_string()));
+                        }
```

Comment:
Once we replace the subject with the VCBC message in the RoundOneJustification, we should replace this logic with a validation of the VCBC message, namely we should check the following:

1. check that the vcbc tag sequence number matches ours
2. check that the vcbc msg tag j is the subject
3. check that it's a c-final message
4. check that the c-final message sig validates against our pub-key-set